conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +317,434 @@ def conda_http_errors(url, repodata_fn):
+        """
+        try:
+            self.load(state_only=True)
+        except FileNotFoundError:
+            self.state.clear()
+        return self.state
+
+    def save(self, data: str):
+        """
+        Write data to <repodata>.json cache path, synchronize state.
+        """
+        temp_path = self.cache_dir / self.cache_dir.with_suffix(f".{os.urandom(4).hex()}.tmp")
+
+        try:
+            with temp_path.open("x") as temp:  # exclusive mode, error if exists
+                temp.write(data)
+
+            return self.replace(temp_path)
+
+        finally:
+            try:
+                temp_path.unlink()
+            except OSError:
+                pass
+
+    def replace(self, temp_path: Path):
+        """
+        Rename path onto <repodata>.json path, synchronize state.
+
+        Relies on path's mtime not changing on move. `temp_path` should be
+        adjacent to `self.cache_path_json` to be on the same filesystem.
+        """
+        with self.cache_path_state.open("a+") as state_file, _lock(state_file):
+            # "a+" avoids trunctating file before we have the lock and creates
+            state_file.seek(0)
+            state_file.truncate()
+            stat = temp_path.stat()
+            # XXX make sure self.state has the correct etag, etc. for temp_path.
+            # UserDict has inscrutable typing, which we ignore
+            self.state["mtime_ns"] = stat.st_mtime_ns  # type: ignore
+            self.state["size"] = stat.st_size  # type: ignore
+            self.state["refresh_ns"] = time.time_ns()  # type: ignore
+            try:
+                temp_path.rename(self.cache_path_json)
+            except FileExistsError:  # Windows
+                self.cache_path_json.unlink()
+                temp_path.rename(self.cache_path_json)
+            state_file.write(json.dumps(dict(self.state), indent=2))
+
+    def refresh(self, refresh_ns=0):
+        """
+        Update access time in .state.json to indicate a HTTP 304 Not Modified response.
+        """
+        with self.cache_path_state.open("a+") as state_file, _lock(state_file):
+            # "a+" avoids trunctating file before we have the lock and creates
+            state_file.seek(0)
+            state_file.truncate()
+            self.state["refresh_ns"] = refresh_ns or time.time_ns()
+            state_file.write(json.dumps(dict(self.state), indent=2))
+
+    def stale(self):
+        """
+        Compare state refresh_ns against cache control header and
+        context.local_repodata_ttl.
+        """
+        if context.local_repodata_ttl > 1:
+            max_age = context.local_repodata_ttl
+        elif context.local_repodata_ttl == 1:
+            max_age = get_cache_control_max_age(self.state.cache_control)
+        else:
+            max_age = 0
+
+        max_age *= 10**9  # nanoseconds
+        now = time.time_ns()
+        refresh = self.state.get("refresh_ns", 0)
+        return (now - refresh) > max_age
```

Comment:
Reuse `timeout` here?

```suggestion
    def stale(self):
        """
        Compare state refresh_ns against cache control header and
        context.local_repodata_ttl.
        """
        return self.timeout() < 0
```

Reply:
timeout loses precision when dividing by `/ 1e9` and is only for a certain debug/log message. Probably not a big deal but it might be better to move or remove the log message.